simpleast-core/src/main/java/com/discord/simpleast/core/parser/Parser.kt

```diff
@@ -29,6 +29,8 @@ open class Parser<R, T : Node<R>> @JvmOverloads constructor(private val enableDe
       remainingParses.add(ParseSpec(null, 0, source.length))
     }
 
+    val filteredRules = if (isNested) rules.filter { it.applyOnNestedParse } else rules
```

Comment:
This could actually be moved to the member-variable level and be computed as rules are added to the parser:

```
fun <C : T> addRule(rule: Rule<R, C>): Parser<R, T> {
     rules.add(rule)
     if (rule.applyOnNestedParse) {
       nestedParseRules.add(rule)
     }
     return this
}
```

kotlin's `filter` allocates a new list and we don't wanna do that in the hot path.

Reply:
Then later on we can do `val resolvedRules = if (isNested) nestedParseRules else rules`

Reply:
good point! updated. :)